rust/src/nasl/builtin/network/udp.rs

```diff
@@ -0,0 +1,115 @@
+use std::{
+    io::{self, Read, Write},
+    net::{IpAddr, SocketAddr, UdpSocket},
+    os::fd::AsRawFd,
+    time::Duration,
+};
+
+use super::{mtu, network_utils::bind_local_socket};
+
+pub struct UdpConnection {
+    socket: UdpSocket,
+    buffer: Vec<u8>,
+    flags: Option<i32>,
+}
+
+const NUM_TIMES_TO_RESEND: usize = 5;
+
+impl Read for UdpConnection {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        for i in 0..NUM_TIMES_TO_RESEND {
+            let result = self.socket.recv_from(buf);
+            match result {
+                Ok((size, origin)) => {
+                    if self.socket.peer_addr()? == origin {
+                        return Ok(size);
+                    }
+                }
+                Err(e) if e.kind() == io::ErrorKind::TimedOut && i != NUM_TIMES_TO_RESEND - 1 => {
+                    self.socket.send(&self.buffer)?;
+                }
+                Err(e) => return Err(e),
+            };
+        }
+        unreachable!()
+    }
+}
+
+impl Write for UdpConnection {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mtu = mtu(self.socket.peer_addr()?.ip());
+        if buf.len() < mtu {
+            let result = unsafe {
+                libc::send(
+                    self.socket.as_raw_fd(),
+                    buf.as_ptr() as *const libc::c_void,
+                    buf.len(),
+                    self.flags.unwrap_or_default(),
+                )
+            };
+            self.flags = None;
+            if result < 0 {
+                return Err(io::Error::last_os_error());
+            }
+            Ok(result as usize)
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "UDP data of size {} exceeds the maximum length of {}",
+                    buf.len(),
+                    mtu
+                ),
+            ))
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
```

Comment:
don't we need flushing?

Reply:
No, as every data is sent out immediately. The internal buffer is for resending that data in case it did not reach its destination on the first try.